src/ch/ch_monitor.c

```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
     return cmd;
 }
 
+static const char *virCHMonitorEventStrings[] = {
+    "vmm:starting",
+    "vmm:shutdown",
+    "vm:booting", "vm:booted",
+    "vm:pausing", "vm:paused",
+    "vm:resuming", "vm:resumed",
+    "vm:snapshotting", "vm:snapshotted",
+    "vm:restoring", "vm:restored",
+    "vm:resizing", "vm:resized",
+    "vm:shutdown", "vm:deleted",
+    "cpu_manager:create_vcpu",
+    "virtio-device:activated", "virtio-device:reset"
+};
+
+static virCHMonitorEvent virCHMonitorEventFromString(const char *sourceStr,
+        const char *eventStr)
+{
+    int i;
+    g_auto(virBuffer) buf = VIR_BUFFER_INITIALIZER;
+    const char *event;
+
+    virBufferAsprintf(&buf, "%s:%s", sourceStr, eventStr);
+    event = virBufferCurrentContent(&buf);
+
+    for (i = 0; i < virCHMonitorEventMax; i++) {
+        if STREQ(event, virCHMonitorEventStrings[i])
+            break;
+    }
+
+    return i;
+}
+
+static int virCHMonitorValidateEventsJSON(char *buffer, int sz,
+            bool *incomplete)
+{
+    int blocks = 0;
+    int events = 0;
+    int i = 0;
+
+    /*
+     * Check if the message is a wellformed JSON.
+     */
+    while (i < sz) {
+        if (buffer[i] == '{') {
+            blocks++;
+        } else if (buffer[i] == '}'){
+            if (blocks == 0) {
+                VIR_WARN("JSON from monitor is not well-formed!");
+                return -1;
+            }
+            blocks--;
+            if (blocks == 0)
+                events++;
+        }
+        i++;
+    }
```

Comment:
I think this is probably the right thing to do for now but this is really dependent on not getting transmission errors from cloud hypervisor. It might be worth extending virjson in the future to handle streaming so we don't need to do this kind of thing.

Reply:
While yajl supports streaming data, it does support only for a single document. In our case each event is one document and we may get multiple documents at the same time. So we would have to manually take care of that I suppose